File: 2022/days/09/main.py
import re
from enum import Enum, auto
from typing import List


class Direction(Enum):
    Up = auto(),
    Down = auto(),
    Left = auto(),
    Right = auto()


def logLocation(tail: List[int], covered: List[List[int]]) -> List[List[int]]:
    if tail not in covered:
        covered.append([tail[0], tail[1]])
    return covered


def strToDirectionEnum(string: str) -> Direction:
    if string == 'U':
        return Direction.Up
    if string == 'D':
        return Direction.Down
    if string == 'L':
        return Direction.Left
    if string == 'R':
        return Direction.Right
    raise Exception(
        f'Could not parse direction, {string} is not a valid direction')


def parseLine(line: str) -> re.Match[str]:
    reCommand = re.compile(r"^ *(\w+) *(\d+) *$")
    matches = reCommand.match(line)
    if matches is None:
        raise Exception(f'Line {line} could not be parsed')
    return matches


def moveRope(rope: List[List[int]], direction: Direction) -> List[List[int]]:
    rope[0] = moveHead(rope[0], direction)
    for i, follower in enumerate(rope[1:]):
        i = i + 1
        leader = rope[i-1]
        follower = tailFollow(leader, follower)

    return rope


def tailFollow(head: List[int], tail: List[int]) -> List[int]:
    # covering
    if tail == head:
        return tail

    xDelta, yDelta = tail[0]-head[0], tail[1]-head[1]

    # in line vertically
    if yDelta == 0:
        if xDelta == - 2:
            tail[0] = head[0] - 1

        if xDelta == 2:
            tail[0] = head[0] + 1

        return tail

    # inline horizontally
    if xDelta == 0:
        if yDelta == - 2:
            tail[1] = head[1] - 1

        if yDelta == 2:
            tail[1] = head[1] + 1

        return tail

    # cover the diagonal cases
    # off by 2,2
    if abs(xDelta) == 2 and abs(yDelta) == 2:
        if xDelta == -2:
            tail[0] = head[0]-1
        else:
            tail[0] = head[0]+1

        if yDelta == -2:
            tail[1] = head[1]-1
        else:
            tail[1] = head[1]+1

        return tail

    # off by 2,1
    # below and offset
    if yDelta == - 2:
        tail[1] = head[1] - 1
        tail[0] = head[0]
        return tail

    # above and offset
    if yDelta == 2:
        tail[1] = head[1] + 1
        tail[0] = head[0]
        return tail

    # left and offset
    if xDelta == - 2:
        tail[0] = head[0] - 1
        tail[1] = head[1]
        return tail

    # right and offset
    if xDelta == 2:
        tail[0] = head[0] + 1
        tail[1] = head[1]
        return tail

    # else they're diagonal one step
    return tail


def moveHead(head, direction):
    if direction == Direction.Up:
        head[1] = head[1] - 1
    if direction == Direction.Down:
        head[1] = head[1] + 1
    if direction == Direction.Left:
        head[0] = head[0]-1
    if direction == Direction.Right:
        head[0] = head[0]+1

    return head


def visualise(rope: List[List[int]]):
    width, height = getRange(rope)

    grid = [['.' for x in range(0, width)] for y in range(0, height)]

    for index, knot in enumerate(rope):
        grid[knot[1]][knot[0]] = str(index)

    grid[rope[0][1]][rope[0][0]] = 'H'
    return grid


def getRange(rope):
    minHorizontal, maxHorizontal = min(
        min([x[0] for x in rope]), 0), max([x[0] for x in rope])
    minVertical, maxVertical = min(
        min([x[1] for x in rope]), 0), max([x[1] for x in rope])
    horizontal = maxHorizontal - minHorizontal + 1
    vertical = maxVertical - minVertical + 1
    return horizontal, vertical


if __name__ == '__main__':
    fileName = 'input.txt'

    head, tail = [0, 0], [0, 0]
    rope = [[0, 0] for i in range(10)]

    coveredPartOne = []
    coveredPartTwo = []
    with open(fileName, 'r') as fp:
        for line in fp:
            line = line.strip('\r\n')
            matches = parseLine(line)
            amount, direction = int(matches[2]), strToDirectionEnum(matches[1])
            while amount > 0:
                head = moveHead(head, direction)
                tail = tailFollow(head, tail)
                rope = moveRope(rope, direction)
                amount = amount-1

                coveredPartOne = logLocation(tail, coveredPartOne)
                coveredPartTwo = logLocation(rope[-1], coveredPartTwo)
                # visual = visualise(rope)
                # for row in visual:
                #     acc = ''
                #     for elem in row:
                #         acc += elem
                #     print(acc)

    print(len(coveredPartOne), len(coveredPartTwo))

File: 2022/days/09/main.ts
import { readFileSync } from 'fs'
import { relative } from 'path'

enum Direction {
    Up = 'U',
    Down = 'D',
    Left = 'L',
    Right = 'R'
}

function getDirection(candidateCharacter: string): Direction {
    switch (candidateCharacter) {
        case 'U':
            return Direction.Up
        case 'D':
            return Direction.Down
        case 'L':
            return Direction.Left
        case 'R':
            return Direction.Right

        default:
            throw new Error(`Unrecognised candidate: ${candidateCharacter}`);
    }
}

function parseLine(line: string): { moveAmount: number, moveDirection: Direction } {
    const reLine = /^(\w) *(\d+) *$/
    let matches = line.match(reLine)
    if (matches) {
        let moveAmount: number = parseInt(matches[2])
        let moveDirection: Direction = getDirection(matches[1])
        return { moveAmount, moveDirection }
    }

    throw new Error(`Cannot parse line correctly:\n${line}`)
}

function moveRope(rope: number[][], moveDirection: Direction): number[][] {

    rope[0] = moveHead(rope[0], moveDirection)
    for (let index = 1; index < rope.length; index++) {
        rope[index] = moveTail(rope[index - 1], rope[index])
    }


    return rope
}

function moveHead(head: number[], moveDirection: Direction): number[] {
    switch (moveDirection) {
        case Direction.Up:
            head[1] -= 1
            break;
        case Direction.Down:
            head[1] += 1
            break;
        case Direction.Left:
            head[0] -= 1
            break;
        case Direction.Right:
            head[0] += 1
            break;
    }
    return head
}

function moveTail(head: number[], tail: number[]): number[] {
    // case head is on top of the tail
    if (head[0] == tail[0] && head[1] == tail[1]) {
        return tail
    }

    let xDelta = tail[0] - head[0]
    let yDelta = tail[1] - head[1]

    // if they're off by one
    if (Math.abs(xDelta) == 1 && Math.abs(yDelta) == 1) {
        return tail
    }

    // in line
    if (xDelta == 0) {
        if (Math.abs(yDelta) == 2) {
            tail[1] = head[1] + yDelta / 2
            return tail
        }
    }
    // in line
    if (yDelta == 0) {
        if (Math.abs(xDelta) == 2) {
            tail[0] = head[0] + xDelta / 2
            return tail
        }
    }

    // two away in both directions
    if (Math.abs(xDelta) == 2 && Math.abs(yDelta) == 2) {
        tail[1] = head[1] + (yDelta / 2)
        tail[0] = head[0] + (xDelta / 2)
        return tail
    }

    // two away in y direction, align in x and move one closer in y
    if (Math.abs(yDelta) == 2) {
        tail[0] = head[0]
        tail[1] = head[1] + (yDelta / 2)
        return tail
    }

    // two away in x, align in y and move one closer in xs
    if (Math.abs(xDelta) == 2) {
        tail[0] = head[0] + (xDelta / 2)
        tail[1] = head[1]
        return tail
    }


    return tail
}


function visualise(rope: number[][]): string[][] {
    function translate(point: number[],minX:number,minY:number): number[] {
        return [point[0] - minX, point[1] -minY]
    }
    const defaultCharacter = '.'
    let xmin = 0
    let xmax = 1
    let ymin = 0
    let ymax = 1

    rope.forEach(knot => {
        xmin = Math.min(xmin, knot[0])
        xmax = Math.max(xmax, knot[0])
        ymin = Math.min(ymin, knot[1])
        ymax = Math.max(ymax, knot[1])
    });

    let height = ymax - ymin + 1
    let width = xmax - xmin + 1

    let head = rope[0]

    let grid: string[][] = Array.from({ length: height }, () => Array.from({ length: width }, () => defaultCharacter));
    for (let index = 0; index < rope.length; index++) {
        const knot = rope[index];
        let translated: number[] = translate(knot,xmin,ymin)
        if (grid[translated[1]][translated[0]] == defaultCharacter) {
            let charToDraw = index.toString()
            if (index == 0) {
                charToDraw = 'H'
            }

            grid[translated[1]][translated[0]] = charToDraw
        }
    }

    return grid
}

function logLocation(visited: number[][], candidate: number[]): number[][] {
    let isNew = true
    for (let index = 0; index < visited.length; index++) {
        const prevLocation = visited[index];
        if (prevLocation[0] == candidate[0] && prevLocation[1] == candidate[1]) {
            isNew = false
            break
        }
    }
    if (isNew) visited.push([...candidate])

    return visited
}

const fileName = 'example.txt'
const file = readFileSync(fileName, 'utf-8');

let rope: number[][] = Array.from({ length: 10 }, () => [0, 0]);
let partOneVistited: number[][] = []
let partTwoVistited: number[][] = []

partOneVistited = logLocation(partOneVistited, rope[1])

file.split('\n').forEach(line => {
    line = line.replace("\r", "")
    if (line.length == 0) return

    let { moveAmount, moveDirection } = parseLine(line)

    while (moveAmount > 0) {
        rope = moveRope(rope, moveDirection)
        partOneVistited = logLocation(partOneVistited, rope[1])
        partTwoVistited = logLocation(partTwoVistited, rope[9])
        moveAmount -= 1
        // console.log(visualise(rope))
    }
});

console.log(partOneVistited.length, partTwoVistited.length)